lib/open_api_spex/schema.ex

```diff
@@ -432,28 +432,42 @@ defmodule OpenApiSpex.Schema do
   def validate(%Schema{type: type}, nil, path, _schemas) do
     {:error, "#{path}: null value where #{type} expected"}
   end
-  def validate(schema = %Schema{type: type}, value, path, _schemas) when type in [:integer, :number] do
+  def validate(schema = %Schema{type: :integer}, value, path, _schemas) when is_integer(value) do
+    # TODO DUPLICATION
+    with :ok <- validate_multiple(schema, value, path),
+         :ok <- validate_maximum(schema, value, path),
+           :ok <- validate_minimum(schema, value, path) do
+      :ok
+    end
+  end
+  def validate(schema = %Schema{type: :number}, value, path, _schemas) when is_number(value) do
+    # TODO DUPLICATION
     with :ok <- validate_multiple(schema, value, path),
          :ok <- validate_maximum(schema, value, path),
          :ok <- validate_minimum(schema, value, path) do
       :ok
     end
   end
-  def validate(schema = %Schema{type: :string}, value, path, _schemas) do
+  def validate(schema = %Schema{type: :string}, value, path, _schemas) when is_binary(value) do
+    # TODO DUPLICATION
     with :ok <- validate_max_length(schema, value, path),
          :ok <- validate_min_length(schema, value, path),
-         :ok <- validate_pattern(schema, value, path),
-         :ok <- validate_enum(schema, value, path) do
+           :ok <- validate_pattern(schema, value, path),
+           :ok <- validate_enum(schema, value, path) do
       :ok
     end
   end
-  def validate(%Schema{type: :boolean}, value, path, _schemas) do
-    case is_boolean(value) do
-      true -> :ok
-      _ -> {:error, "#{path}: Invalid boolean: #{inspect(value)}"}
+  def validate(schema = %Schema{type: :string}, value = %DateTime{}, path, _schemas) do
+    # TODO DUPLICATION
+    with :ok <- validate_max_length(schema, value, path),
```

Comment:
I think any non-nil value for `maxLength`  /  `minLength` / `pattern` here will raise an `ArgumentError` when given a `%DateTime{}` struct?

I'd expect something like:

```elixir
def validate(schema = %Schema{type: :string, format: :"date-time"}, value = %DateTime{}, _path, _schemas) do
 :ok
end
```

Reply:
Looks like it's also causing a dialyzer failure since those validation functions are `@spec`ed to accept only `String.t`